### ISSUE TITLE
Automatic Server Content Delivery through Git

### ISSUE THREAD

**Issue by bizzclaw** [Facepunch/garrysmod-requests#1030]
# Current Delivery Methods
Okay, so right now we have two options for automatically sending content to players:

### Source Downloading 
Source has a built in system for downloading content, but only if the user doesn't already have the content. This means that if assets are updated, the user has to manually remove old versions before downloading the new versions

### resource.AddWorkshop 
Allows server owners to indicate workshop addons a user will automatically subscribe too, but only for that server. This is good, because addons will re-download whenever they're updated, but this is also a detriment, because the entire addon has to be downloaded, even if the update only applies to two or three files. Also, the maximum size of workshop addons is also a point of contention because it's a pain to have to manage 10+ addons for one server from the workshop, and it's an eyesore for people browsing the workshop to constantly try to sift through ServerContent addons.

# Proposal
While both of the above methods **work**, they have quite a few issues associated with them, but maybe there's another way. A way anyone who's browsing this site is already familiar with

### Git
The set-up could be simple, server-owners could host their content on GitLab which has a [10GB repository size cap](https://about.gitlab.com/2015/04/08/gitlab-dot-com-storage-limit-raised-to-10gb-per-repo/), or host their own git server. Then, they specify what git repositories / branch players will pull from whenever they join the server, and while a player joins, they will automatically _pull_ the latest version from that repository/branch.

The same filter used for uploading workshop addons can be applied to players who download using this git to ensure there's no fowl play, 

This would in turn allow players to download updates to specific files without re-downloading the whole pack, Server Owners won't have to make and manage a series of addons specific to a server, which would reduce the rate at which ServerContent addons are flooding the workshop.

Since [Git is open source](https://git-scm.com) there would be no licensing required to implement it, furthermore, cross platform libraries [already exist for implementing git in to C++ and other languages](https://libgit2.github.com), heck, even [Lua](https://github.com/libgit2/luagit2) for some reason. So such an implementation isn't impossible.

Anyways, I hope this gets some consideration because while some may see it as silly to have servers with gigabytes of content, Garrysmod is a unique platform that is incredibly easy to build upon, and is built on an engine that can make use of all of that content. I'd say it's a waste not to take advantage of that support, and a reliable and easy method for getting that content to players will make a huge difference.

# TL;DR:
Please add automatic git downloading for servers  because the current methods are unstable, unreliable and a headache in general.

**Comment by stickz** [comment]
This will get your GitHub account suspended abusing it as a web server. LOL

**Comment by bizzclaw** [comment]
Github isn't the only place to host git repositories. Setting up your own git server is easy. Also, no matter how much you compress your textures, if you have a massive quantity of content, your going to hit those limits.

**Comment by stickz** [comment]
bizzclaw, I'm not talking about compression techniques. I'm talking about reducing the size of the content BEFORE it's compressed, by adjusting the size of the .vtf image file. 

You can average less than 500kb per player model compressed - doing this and never hit the limits.

**Comment by bizzclaw** [comment]
First of all I think you mean 1024x1024 and 512x512.
Secondly there's more to gamemode content than "playermodels".

**Comment by SamuelMoriarty** [comment]
How different would this be from the old FastDL system? I don't see much benefit for using git over FastDL in this case, apart from maybe having git compress your shit for you which is faster for a lot of small files.

Either way, you can implement a similar system entirely yourself with 2 components:
clientside - simple Lua code that receives a list of hashes for each file and tries to find those files locally in data/, then makes a request with all missing/outdated files to the server with HTTP()
serverside - a binary module (or custom app server) that sends the files to the client in a compressed format over HTTP

After the client has all the raw files in their data/ folder you can build a .gma archive on the fly and mount it. Skip building the .gma if no changes are present and all the hashes are the same.

Sure you could do this all much more efficiently natively in the engine, but the aforementioned method should work fairly well in practice.

**Comment by stickz** [comment]
You can use this client-side function to cache a workshop mod, while a client is playing. This is insanely useful for maps. If used correctly with workshop, it would remove all but the first map download.

This will indirectly curve part of the issue, of clients having to constantly re-download things.
```
function DownloadWorkshopFile(workshop_mod_id)
       steamworks.FileInfo( workshop_mod_id, function( result )
		if result.installed == false and not steamworks.IsSubscribed(workshop_mod_id) then
			steamworks.Download( result.fileid, false, function( name )			
				chat.AddText(Color(196, 187, 143), "Cached workshop mod on hard-drive")
			end )
		end	
        end)
end
```

A simple if-statement on the server side can be used to mount a map only when being played.
```
--Scripts for mounting the map packs
local currentMap = game.GetMap()

-- If the current map is on a seperate list
if currentMap == "mor_facility_cv2" then
	resource.AddWorkshop("811812938") -- mor_facility_cv2
	
elseif currentMap == "mor_chemical_labs_b3_re" then
	resource.AddWorkshop("658240507") -- mor_chemical_labs_b3_re
end
```

**Comment by SupinePandora43** [comment]
> ```
> --Scripts for mounting the map packs
> local currentMap = game.GetMap()
> 
> -- If the current map is on a seperate list
> if currentMap == "mor_facility_cv2" then
> 	resource.AddWorkshop("811812938") -- mor_facility_cv2
> 	
> elseif currentMap == "mor_chemical_labs_b3_re" then
> 	resource.AddWorkshop("658240507") -- mor_chemical_labs_b3_re
> end
> ```
this is very cool and optimized way for delivering maps to client!
but most of servers just use ONE BIG COLLECTION with millions of addons.
###### WoW! why when i play on friend's server (p2p), it uses this, but when i play on metrostroi server it don't :C
>Either way, you can implement a similar system entirely yourself with 2 components:
>clientside - simple Lua code that receives a list of hashes for each file and tries to find those files locally in data/, then makes a request with all missing/outdated files to the server with HTTP()
>serverside - a binary module (or custom app server) that sends the requested files to the client in a compressed format over HTTP

Server-side
```
util.AddNetworkString("git-addons")
hook.Add( "PlayerInitialSpawn", "some_unique_name", function(ply) 
timer.Simple(10, function()
net.Start("git-addons")
net.WriteString(GetConVar("sv_repository"))
net.Send(ply)
end)
end)
```
client:
```
LocalPlayer():ChatPrint( "wait until your client receive data from server" )
net.Receive( "git-addons", function( len, ply )
local repo = net.ReadString()
if repo ~= "" then
\# git
\# compile to gma
game.MountGMA(PATH TO GMA)
end
end)
```

// i will create addon on this idea, sorry but i'm in school, and i can't write all code

**Comment by robotboy655** [comment]
For the last time, you don't need to call `resource.AddWorkshop` on currently active map, it is done FOR YOU AUTOMATICALLY as long as the map is in the collection.

**Comment by Kefta** [comment]
You can setup your own local git mirror that also acts as a FastDL webserver. I am fairly certain that GitLab doesn't want you using its repository hosting as a content distribution platform.

**Comment by bizzclaw** [comment]
Wow this is ancient. I started this suggestion years ago before I knew a lot more about git than I do now. I based this idea off the times when everyone downloaded PHX, Wire and other server content through tortoise svn and while I still think full version control for content would be ideal, git is kind of overkill for that and the new workshop's increased limits plus functionality to only make people download the changed files is more than sufficient. There's probably much more important work to be done.
